Windows facts for licensing information

rhydian76 opened the pull request:
Windows facts for licensing information added

##### SUMMARY
No available facts for windows licensing information - for things like activation status, license channel, and license edition.

##### ISSUE TYPE

- Feature Pull Request


##### COMPONENT NAME

/lib/ansible/modules/windows/win_product_facts.ps1
/lib/ansible/modules/windows/win_product_facts.py

##### ADDITIONAL INFORMATION

The win_product_facts module now adds facts for ansible_os_license_edition, ansible_os_license_channel and ansible_os_license_status

rhydianjenkinscs committed:
Windows facts for ansible_winlicense_edition, ansible_winlicense_channel, ansible_winlicense_status added

rhydianjenkinscs committed:
bugfix, replaced = with :

dagwieers commented on lib/ansible/modules/windows/win_product_facts.py:
```diff
@@ -33,4 +33,7 @@
   sample:
     ansible_os_product_id: 00326-10000-00000-AA698
     ansible_os_product_key: T49TD-6VFBW-VV7HY-B2PXY-MY47H
+    ansible_winlicense_edition: Windows(R) ServerStandard edition
```
Why are we naming this `winlicense`?

I would rather go for `ansible_os_license_edition`, `ansible_os_license_channel` and `ansible_os_license_status`.

rhydian76 replied:
OK thats fine. Ill do that now and commit

dagwieers commented on lib/ansible/modules/windows/win_product_facts.py:
```diff
@@ -11,13 +11,13 @@
 ---
 module: win_product_facts
-short_description: Provides Windows product information (product id, product key)
+short_description: Provides Windows product information (product id, product key, license status, license channel, license edition)
```
We probably don't want to make the *short* description that long.
Maybe we can now change this to:
```yaml
short_description: Provide Windows product and license information
```

rhydian76 replied:
done.

dagwieers commented on lib/ansible/modules/windows/win_product_facts.py:
```diff
@@ -11,13 +11,13 @@
-options: {}
 author:
 - Dag Wieers (@dagwieers)
+options: {}
```
Please leave the author at the end (like the generated documentation).

rhydian76 replied:
Done.

rhydianjenkinscs committed:
changed facts to ansible_os_license_* in win_product_facts.ps1
Shortened the short_description in win_product_facts.py
removed errant - options: {} from  win_product_facts.py

dagwieers reviewed (approved):
LGTM

jborean93 commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,59 @@ if (-not $product_key) {
+try {
+    $license_info = Get-CimInstance SoftwareLicensingProduct | Where-Object PartialProductKey
+
+    switch ($license_info.LicenseStatus){
```
You don't have to change it but this can easily be simplified with

```
$winlicense_status = switch ($license_info.LicenseStatus) {
    0 { "Unlicensed" }
    1 { "Licensed" }
    2 { "OOBGrace" }
    ...
    default { "NA" }
}
```

The benefits of this approach is that you only need to set the var name once meaning less complication.

jborean93 commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,59 @@ if (-not $product_key) {
     }
 }
 
+try {
```
Why do you have this in a try/catch? I'm able to retrieve this object even under a non-admin process so I don't think we need this. It may be best to get the CIM object then set the vars to NA if the property we are trying to get is null.

dagwieers replied:
Personally I prefer to have a value of $null (null) instead.

rhydian76 replied:
I used NA as I have seen it in use elswhere. But I can change it to $null, thats no problem.

The try/catch was put in to protect against errors caused by some weird behaviour with antivirus blocking some of the powershell calls and classing them as malicious (BitDefender Hyper Detect).

jborean93 commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,59 @@ if (-not $product_key) {
+        }
+
+        default {
+            $winlicense_status="NA"
```
I feel this null or unexpected values should be N/A not NA. NA could potentially be an acronym for something else whereas N/A is the standard I've seen for not applicable.

dagwieers replied:
I'd got for $null, rather than a string. Unless that string somehow is the value that is expected in this case.

jborean93 commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,59 @@ if (-not $product_key) {
+    $winlicense_channel = $license_info.ProductKeyChannel
+
+} catch {
+    $winlicense_edition = "NA"
```
Same comments for these vars around NA vs N/A.

dagwieers replied:
null ;-)

rhydian76 committed:
Code optimisations and changed return code to $null from NA when unknown license information returned from CIM SoftwareLicensingProduct class.

rhydian76 commented:
I have committed the requested changes.

dagwieers commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,41 @@ if (-not $product_key) {
+    default { $null }
+}
+
+if ($license_info.Name) {
```
Can we not just do ?
```powershell
$winlicense_edition = $license_info.Name
```

dagwieers commented on lib/ansible/modules/windows/win_product_facts.ps1:
```diff
@@ -57,9 +57,41 @@ if (-not $product_key) {
+    $winlicense_edition = $null
+}
+
+if ($license_info.ProductKeyChannel) {
```
Can we not just do?
```powershell
$winlicense_channel = $license_info.ProductKeyChannel
```

rhydian76 replied:
Yes, I guess so. - I havent had my coffee this morning! :)
Committed these changes

rhydian76 committed:
code optimisation

rhydian76 reviewed (commented):
All requested changes have been made

dagwieers commented:
@rhydian76 I am happy with this, would like to use it too.

dagwieers commented:
@jborean93 Are we ready to land?

jborean93 commented:
@dag if you are happy I'm happy.

jhawkesworth commented:
Tested against windows 10, so I think this can be merged.

```
TENSY | SUCCESS => {
    "ansible_facts": {
        "ansible_os_license_channel": "Volume:MAK",
        "ansible_os_license_edition": "Windows(R), Enterprise edition",
        "ansible_os_license_status": "Licensed",
        "ansible_os_product_id": "SOME NUMBERS WERE HERE",
        "ansible_os_product_key": "A LICENCE KEY WAS HERE"
    },
    "changed": false
}
```

jborean93 commented:
Thanks everyone for testing this one out.
